cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrader.java

```diff
@@ -0,0 +1,320 @@
+                        getControllerStateByStatus(curCompletedRunRecord.getStatus()));
+        }
+        return null;
+      }
+    });
+  }
+
+  /**
+   * Writes the {@link AppMetadataStore#TYPE_RUN_RECORD_STARTED} entry in the app meta table so that
+   * {@link AppMetadataStore#TYPE_RUN_RECORD_COMPLETED} can be written which deleted the started record.
+   *
+   * @param appId the application id
+   * @param programType {@link ProgramType} of this program
+   * @param programId the program id of this program
+   * @param pId the process id of the run
+   * @param startTs the startTs
+   */
+  private void writeTempRunRecordStart(String appId, ProgramType programType, String programId, String pId,
+                                       long startTs) {
+    store.setStart(Id.Program.from(Id.Application.from(Constants.DEFAULT_NAMESPACE, appId), programType, programId),
+                   pId, Long.MAX_VALUE - startTs);
+  }
+
+  /**
+   * Writes the updated application metadata through the {@link DefaultStore}
+   *
+   * @param appMeta the {@link ApplicationMeta} of the app
+   * @throws URISyntaxException if failed to create {@link URI} from the archive location in metadata
+   */
+  private void upgradeAppMeta(ApplicationMeta appMeta) throws URISyntaxException, IOException {
+    store.addApplication(Id.Application.from(Constants.DEFAULT_NAMESPACE, appMeta.getId()), appMeta.getSpec(),
+                         updateAppArchiveLocation(appMeta.getId(), new URI(appMeta.getArchiveLocation())));
+    // store the stream names which belongs to this app as we want to filter out streams which do not belong to any app
+    // later
+    appStreams.addAll(appMeta.getSpec().getStreams().keySet());
+  }
+
+  /**
+   * Creates the new archive location for application with namespace
+   *
+   * @param appId : the application id
+   * @param archiveLocation : the application archive location
+   * @return {@link Location} for the new archive location with namespace
+   */
+  private Location updateAppArchiveLocation(String appId, URI archiveLocation) throws IOException {
+    String archiveFilename = locationFactory.create(archiveLocation).getName();
+
+    return locationFactory.create(Constants.DEFAULT_NAMESPACE).append(
+      cConf.get(Constants.AppFabric.OUTPUT_DIR)).append(appId).append(Constants.ARCHIVE_DIR).append(archiveFilename);
+  }
+
+  /**
+   * Gives the {@link ProgramController.State} for a given {@link ProgramRunStatus}
+   * Note: This will given the state as {@link ProgramController.State#STARTING} for {@link ProgramRunStatus#RUNNING}
+   * even though running has multiple mapping but that is fine in our case as we use this
+   * {@link ProgramController.State} to write the runRecordCompleted through {@link DefaultStore#setStop} which converts
+   * it back to {@link ProgramRunStatus#RUNNING}. So we don't really care about the temporary intermediate
+   * {@link ProgramController.State}
```

Comment:
i see... this is bad though because we only know this by looking at the implementation of AppMetadataStore.  I think instead of this we can refactor the api to take a ProgramRunStatus, since that is what is being used anyway.